common/persistence/cassandra/queue_v2_store.go

```diff
@@ -233,27 +244,82 @@ func (q *queueV2Store) CreateQueue(
 	return &persistence.InternalCreateQueueResponse{}, nil
 }
 
-func (q *queueV2Store) getMinMessageID(
+func (s *queueV2Store) RangeDeleteMessages(
+	ctx context.Context,
+	request *persistence.InternalRangeDeleteMessagesRequest,
+) (*persistence.InternalRangeDeleteMessagesResponse, error) {
+	if request.InclusiveMaxMessageMetadata.ID < persistence.FirstQueueMessageID {
+		return nil, fmt.Errorf(
+			"%w: id is %d but must be >= %d",
+			persistence.ErrInvalidQueueRangeDeleteMaxMessageID,
+			request.InclusiveMaxMessageMetadata.ID,
+			persistence.FirstQueueMessageID,
+		)
+	}
+	q, err := s.getQueue(ctx, request.QueueType, request.QueueName)
+	if err != nil {
+		return nil, err
+	}
+	minMessageID := s.getPartitionMinMessageID(q.Metadata, request.QueueType, request.QueueName)
+	maxID := request.InclusiveMaxMessageMetadata.ID
+	err = s.session.Query(
+		TemplateRangeDeleteMessagesQuery,
+		request.QueueType,
+		request.QueueName,
+		0, // partition
+		minMessageID,
+		maxID,
+	).WithContext(ctx).Exec()
+	if err != nil {
+		return nil, gocql.ConvertError("QueueV2RangeDeleteMessages", err)
+	}
+	partitions := q.Metadata.Partitions
+	if len(partitions) != 1 {
+		partitionWarningMsg := s.getPartitionWarningMsg(
+			request.QueueType,
+			request.QueueName,
+			len(partitions),
+		)
+		s.logger.Warn(fmt.Sprintf(
+			"%s This query to delete messages assumes that there is only 1 partition. "+
+				"The messages are deleted, but we are not updating the min message ID of any partition.",
+			partitionWarningMsg,
+		))
+		return &persistence.InternalRangeDeleteMessagesResponse{}, nil
+	}
+	partitions[0].MinMessageId = maxID + 1
+	bytes, _ := q.Metadata.Marshal()
+	applied, err := s.session.Query(
+		TemplateUpdateQueueMetadataQuery,
+		bytes,
+		enums.ENCODING_TYPE_PROTO3.String(),
+		q.Version+1,
+		request.QueueType,
+		request.QueueName,
+		q.Version,
+	).WithContext(ctx).MapScanCAS(make(map[string]interface{}))
+	if err != nil {
+		return nil, gocql.ConvertError("QueueV2UpdateQueueMetadata", err)
+	}
+	if !applied {
+		return nil, fmt.Errorf(
```

Comment:
When there're two concurrent delete requests. Say request 1 wants to delete 1 to 10 and request 2 wants to delete 1 to 20. Request 2 succeeds and request 1 got a conflict. They request 1 will retry, and overwrite the ack level from 20 to 10? I think we need to validate min & max message ID before serving the request.

Reply:
So, I spent a long time thinking about this, and I don't know if there's any solution with Cassandra. It's ok if the `min_message_id` is less than it should be: that will just hurt read performance slightly until it is fixed by the next delete that doesn't have a conflict. In any case, it's always possible for `min_message_id` to be too small because we could crash after deleting and before writing the `min_message_id` back. I think it's basically like this:

On enqueue, we have to write the next_message_id first, then enqueue the message (that means we might skip message ids, but that's ok--they just have to increase monotonically). On dequeue (delete), we have to delete the messages and then write the min_message_id (if we write it first, we could crash before we actually delete the messages, and they'll become zombies).

Reply:
synced offline: There is a way to fix it:
```
1. Load queue metadata (min_msg_id, version)
2. If min_msg_id > request.max_msg_id, return success
3. Delete messages
4. CAS to update min_msg_id, and return Unavailable if !applied
```
Then, clients can retry the `Unavailable` error.

Reply:
This is now done, and there's a test which verifies that, regardless of whether the smaller delete query succeeds first or not, after the failing query is retried, the min_message_id is correct.